src/qpx_mmc.cpp

```diff
@@ -0,0 +1,4244 @@
+        drive->media.gbpl = 33;
+
+      read_mediaid_bd(drive);
+      if (!drive->silent)
+        printf("** MID: '%s'\n", drive->media.MID);
+    }
+  } else {
+    read_capacity(drive);
+    if (drive->media.capacity) {
+      drive->media.type = DISC_CDROM;
+      read_disc_information(drive);
+    }
+    return 0;
+  }
+  return 1;
+}
+
+int get_spindown(drive_info *drive) {
+  mode_sense(drive, 0x0D, 00, 192);
+  if (drive->err) {
+    drive->parms.spindown_idx = spindowns;
+    return (drive->err);
+  }
+  drive->parms.spindown_idx = drive->rd_buf[11] & 0x0F;
+  return 0;
+}
+
+int set_spindown(drive_info *drive) {
+  memset(drive->rd_buf, 0, 16);
+  drive->rd_buf[8] = 0x0D;
+  drive->rd_buf[9] = 0x06;
+  drive->rd_buf[11] = drive->parms.spindown_idx & 0x0F;
+  drive->rd_buf[13] = 0x3C;
+  drive->rd_buf[15] = 0x4B;
+  mode_select(drive, 16);
+  return 0;
+}
+
+int get_performance(drive_info *drive, bool rw, uint8_t type) {
+  const int max_descs = 52;
+  const int desc_len = 16;
+  uint32_t len, descn;
+  uint32_t r, w, lba;
+  //	int	i;
+  int j, offs;
+  drive->cmd[0] = MMC_GET_PERFORMANCE;
+  drive->cmd[1] = (!type) ? (0x04 * rw) : 0x00;
+  drive->cmd[8] = (max_descs >> 8) & 0xFF;
+  drive->cmd[9] = max_descs & 0xFF;
+  drive->cmd[10] = type;
+  drive->cmd[11] = 0x00;
+  if ((drive->err = drive->cmd.transport(READ, drive->rd_buf, 1024))) {
+    if (!drive->silent)
+      sperror("GET_PERFORMANCE", drive->err);
+    return (drive->err);
+  }
+  //	return 1;
+  len = ntoh32(drive->rd_buf);
+  descn = len / desc_len;
+  //	printf("Performance data length: %d; decriptors: %d\n",len, descn);
+
+  //	printf("GET_PERFORMANCE response dump:\n");
+  //	for (j=0; j*8<len; j++) {
+  //		for (i=0; i<8; i++) printf(" 0x%02X",drive->rd_buf[i]);
+  //		printf("\n");
+  //	}
+
+  if (type == 0x03)
+    for (j = 0; j < speed_tbl_size; j++)
+      drive->parms.wr_speed_tbl_kb[j] = -1;
+
+  for (j = 0; j < min(descn, (type == 0x03) ? speed_tbl_size : descn); j++) {
+    offs = 8 + j * desc_len;
+    //		printf("\t%s descriptor #%02d:", rw ? "Write" : "Read",j);
+
+    if (!type) {
+      drive->perf.lba_s = ntoh32(drive->rd_buf + offs);
+      offs = 8 + j * desc_len + 4;
+      drive->perf.spd_s = ntoh32(drive->rd_buf + offs);
+      offs = 8 + j * desc_len + 8;
+      drive->perf.lba_e = ntoh32(drive->rd_buf + offs);
+      offs = 8 + j * desc_len + 12;
+      drive->perf.spd_e = ntoh32(drive->rd_buf + offs);
+
+      offs = 8 + j * desc_len + 12;
+      if (!rw)
+        drive->parms.read_speed_kb = ntoh32(drive->rd_buf + offs);
+      else
+        drive->parms.write_speed_kb = ntoh32(drive->rd_buf + offs);
+
+      //			printf("\t%dkB/s@%d -> %dkB/s@%d\n",
+      //				drive->perf.spd_s,drive->perf.lba_s,drive->perf.spd_e,drive->perf.lba_e);
+    } else if (type == 0x03) {
+      offs = 8 + j * desc_len + 4;
+      lba = ntoh32(drive->rd_buf + offs);
+      offs = 8 + j * desc_len + 8;
+      r = ntoh32(drive->rd_buf + offs);
```

Comment:
Codacy found an issue: [Variable 'r' is assigned a value that is never used.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)